README.md
# n-assert
Node.js assertion library.

[![Build Status](https://github.com/AlexanderMac/n-assert/workflows/CI/badge.svg)](https://github.com/AlexanderMac/express-starter/actions?query=workflow%3ACI)
[![Code Coverage](https://codecov.io/gh/AlexanderMac/n-assert/branch/master/graph/badge.svg)](https://codecov.io/gh/AlexanderMac/n-assert)
[![npm version](https://badge.fury.io/js/n-assert.svg)](https://badge.fury.io/js/n-assert)

### Why
To assert the complex objects.

```js
// Some test method returns this object
let actual = {
  _id: '5945bf36ccb3fa0011e8533c',
  name: 'John',
  email: 'john@mail.com',
  phones: [
    { type: 'mobile', number: '1234567' }
    { type: 'work', number: '567382' }
  ],
  account: {
    number: '11111',
    registered: '2010-04-21'
  },
  createdAt: '2017-02-03'
}

// Ned to do the following assertions:
//   _id is exists
//   name and email match
//   mobile phone is exists and it's a number
//   account registered match
let expected = {
  _id: '_mock_',
  name: 'John',
  email: 'john@mail.com',
  phones: [
    { type: 'mobile', number: /^d{1,10}$/ }
  ],
  account: {
    registered: '2010-04-21'
  }
}

nassert.assert(actual, expected)
```

### Install
```bash
$ npm i -D n-assert
```

### Usage
```js
const nassert = require('n-assert')

should('should find user by name', async () => {
  let actual = await usersSrvc.getUserByName('John')
  let expected = { /* some data */ }

  nassert.assert(actual, expected)
})
```

### API
- **assert(actual, expected, isEqual)**<br>
Asserts that `actual` is equal to `expected`.

  - `actual` - actual object.
  - `expected` - expected object.
  - `isEqual` - performs a deep assertion between `actual` and `expected` to determine that they are indentical. 

- **assertResponse(res, expectedStatus, expectedBody)**<br>
Asserts that status and body in `res` is equal to `expectedStatus` and `expectedBody`.

  - `res` - http response.
  - `expectedStatus` - expected http response status.
  - `expectedBody` - expected http response body.

- **assertFn**<br>
Asserts that stubbed function has been called or not, and if called, then with provided arguments. **Warning!** To use `sinonMatch` function, sinon instance should be initialized via `nassert.initSinon(sinon)` method.

  - `inst` - object instance.
  - `fnName` - stubbed function name in the service.
  - `callCount` - count of function calles, one by default.
  - `nCall` - validates that method was called with provided args on that call, zero by default.
  - `expectedArgs` - expected _single_ argument, should be `__without-args__` if function is called without arguments.
  - `expectedMultipleArgs` - expected _few_ arguments.

- **sinonMatch(expected)**<br>
Calls sinon.match and compares `actual` value with `expected` using `nassert.assert`. Returns `true` if values identical, otherwise throws an error. **Warning!** To use `sinonMatch` function, sinon instance should be initialized via `nassert.initSinon(sinon)` method.

- **getObjectId()**<br>
Returns new mongodb ObjectId.

- **getObjectIdStr()**<br>
Returns new mongodb ObjectId in string format.

### License
Licensed under the MIT license.

### Author
Alexander Mac


CHANGELOG.md
# <sub>v7.0.0</sub>
#### _Feb. 26, 2021_
  * Drop support for Node versions prior to v12.
  * Update dependencies.

# <sub>v6.0.0</sub>
#### _Jan. 24, 2020_
  * Drop support for Node v6.
  * Update dependencies.

# <sub>v5.0.2</sub>
#### _Apr. 5, 2019_
  * Remove mongoose methods. To generate and validate ObjectId, bson package has been installed.
  * Method `assertCollection` moved to a new package [mongo-assert](https://github.com/AlexanderMac/mongo-assert).
  * Remove `processError` method.
  * Remove `resolveOrReject` method.
  * Rename `validateCalledFn` method to `assertFn`, change `srvc` parameter to `inst`.
  * Update dependencies.

# <sub>v4.0.0</sub>
#### _Nov. 15, 2018_
  * Remove `sinon` dependency, now it should be passed via `nassert.initSinon` method.

# <sub>v3.1.0</sub>
#### _Aug. 4, 2018_
  * Add `getObjectIdStr` method, that returns `mongoose.ObjectId` in string format.
  * Add ability to use `validateCalledFn` method to assert multiple calls. Add `callCount` and `nCall` parameters.

# <sub>v3.0.0</sub>
#### _Jan. 27, 2018_
  * Update sinon to v4.x.
  * Update mocha to v5.x, mongoose to v5.x, should to v13.x.
  * Remove unused Istanbul dependecy.

# <sub>v2.2.0</sub>
#### _Jan. 6, 2018_
  * Add `validateCalledFn` method to validate that stubbed function is called or not, and if called then with provided arguments.

# <sub>v2.1.0</sub>
#### _Oct. 2, 2017_
  * Add `isEqual` parameter to `assert` method to perform a deep assertion between two values to determine if they are equivalent.
  * Add logic for converting `actual` mongoose documents array to plain objects.

# <sub>v2.0.0</sub>
#### _Sep. 1, 2017_
  * Delete deprecated methods: `getList`, `getSingleById`, `sinonMatch`, `buildQuery`, `processErrorNoMessage`.
  * Some methods now are private: `assertId`, `isSimplePrim`.
  * Remove `getObjectIdStr` method.
  * Add `assertCollection` method. 
  * Add `safeToString` method to safety convert value to string.
  * Add parameter validations in `assertCollection` method.
  * Improve `assert` method algorithm, decreased the number of recursive calls.
  * AssertionError now contains `at path` part (contains the wrong value full path).
  * Fix bug in `assert` method for the case when an array of objects is asserting.
  * Fix a few bugs in assertions.
  * Update dependencies.

# <sub>v1.1.0</sub>
#### _Aug. 20, 2017_
  * Cover codebase by tests with code coverage ~99%.

# <sub>v1.0.0</sub>
#### _Aug. 19, 2017_
  * Release a stable version.
  * A few major changes in used dev-tools (eslint instead of jshint, add mocha for test and istanbul for code coverage).
 
# <sub>v0.1.0</sub>
#### _Jun. 8, 2017_
  * Revert sinon from v2 to v1.

# <sub>v0.0.1</sub>
#### _Jun. 8, 2017_
  * First release.


test/test.js
'use strict';

let should  = require('should');
let nassert = require('../index');

describe('test', () => {
  it('should', () => {
    nassert.getObjectId();
    should(10).equal(10);
  });
});


test/index.js
const _ = require('lodash')
const should = require('should')
const sinon = require('sinon')
const nassert = require('../index')

nassert.initSinon(sinon)

describe('n-assert', () => {
  describe('assert', () => {
    function test(actual, expected, isEqual, expectedErr) {
      try {
        nassert.assert(actual, expected, isEqual)
        if (expectedErr) {
          throw new Error('Test must fail, because an error is expected')
        }
      } catch (err) {
        if (expectedErr instanceof Error) {
          should(err.message).eql(expectedErr.message)
        } else {
          throw err
        }
      }
    }

    describe.skip('actual is Mongoose document', () => {
      function User() {
        // TODO: implement
      }

      it('should convert a single actual Mongoose document to a plain object', () => {
        let actual = new User({
          name: 'John',
          email: 'john@mail.com'
        })
        let expected = {
          name: 'John',
          email: 'john@mail.com'
        }
        test(actual, expected)
      })

      it('should convert an array of actual Mongoose document to plain objects', () => {
        let actual = [
          new User({ name: 'John', email: 'john@mail.com' }),
          new User({ name: 'Donald', email: 'donald@mail.com' })
        ]
        let expected = [
          { name: 'John', email: 'john@mail.com' }
        ]
        test(actual, expected)
      })
    })

    describe('isEqual is passed', () => {
      it('should fail when isEqual is passed and expected not equal to actual', () => {
        let actual = [
          { name: 'John', email: 'john@mail.com' },
          { name: 'Donald', email: 'donald@mail.com' }
        ]
        let expected = [
          { name: 'John', email: 'john@mail.com' }
        ]
        /* eslint quotes: off */
        let expectedErr = new Error(`expected Array [ '0.email', '0.name', '1.email', '1.name' ] to equal Array [ '0.email', '0.name' ] (at length, A has 4 and B has 2)`)
        test(actual, expected, true, expectedErr)
      })

      it('should pass when isEqual is passed and expected equals to actual', () => {
        let actual = [
          { name: 'John', email: 'john@mail.com' },
          { name: 'Donald', email: 'donald@mail.com' }
        ]
        let expected = [
          { name: 'John', email: 'john@mail.com' },
          { name: 'Donald', email: 'donald@mail.com' }
        ]
        test(actual, expected, true)
      })
    })

    describe('expected is nil', () => {
      it('should fail assertion when actual is not undefined', () => {
        let actual = 1
        let expected = undefined
        let expectedErr = new Error('expected 1 not to be truthy (false negative fail)')
        test(actual, expected, undefined, expectedErr)
      })

      it('should pass assertion when actual is undefined and expected is null', () => {
        let actual = undefined
        let expected = null
        test(actual, expected)
      })

      it('should pass assertion when actual is undefined and expected is undefined', () => {
        let actual = undefined
        let expected = undefined
        test(actual, expected)
      })
    })

    describe('expected is an array of simple primitives', () => {
      it('should fail assertion when actual is not an array', () => {
        let actual = 1
        let expected = [1, 2, 3]
        let expectedErr = new Error('expected 1 to equal Array [ 1, 2, 3 ]')
        test(actual, expected, undefined, expectedErr)
      })

      it('should fail assertion when actual is not equal to expected', () => {
        let actual = [1, 2, 3]
        let expected = [1, 'a', 3]
        let expectedErr = new Error('expected Array [ 1, 2, 3 ] to equal Array [ 1, \'a\', 3 ] (at \'1\', A has 2 and B has \'a\')')
        test(actual, expected, undefined, expectedErr)
      })

      it('should pass assertion when actual is equal to expected', () => {
        let actual = [1, 2, 3]
        let expected = [1, 2, 3]
        test(actual, expected)
      })
    })

    describe('expected is a simple primitive', () => {
      it('should fail assertion when actual is not simple primitive', () => {
        let actual = { name: 'John' }
        let expected = 5
        let expectedErr = new Error('expected Object { name: \'John\' } to equal 5')
        test(actual, expected, undefined, expectedErr)
      })

      it('should fail assertion when actual is not equal to expected', () => {
        let actual = 1
        let expected = 5
        let expectedErr = new Error('expected 1 to equal 5')
        test(actual, expected, undefined, expectedErr)
      })

      it('should pass assertion when actual is equal to expected', () => {
        let actual = 5
        let expected = 5
        test(actual, expected)
      })
    })

    describe('expected is a complex object', () => {
      let getDefActual = (ex) => {
        let def = {
          _id: nassert.getObjectId(),
          __v: 2,
          createdAt: new Date(2017, 15, 20),
          updatedAt: new Date(2017, 15, 21),
          name: 'John Smith'
        }
        return _.extend(def, ex)
      }

      it('should pass assertion when actual is equal to expected with <string>_id', () => {
        let objId = nassert.getObjectId()
        let actual = getDefActual({ _id: objId })
        let expected = {
          _id: objId.toString()
        }
        test(actual, expected)
      })

      it('should pass assertion when actual is equal to expected with <ObjectId>_id', () => {
        let objId = nassert.getObjectId()
        let actual = getDefActual({ _id: objId })
        let expected = {
          _id: objId
        }
        test(actual, expected)
      })

      it('should pass assertion when actual is equal to expected with <_mock_>_id', () => {
        let objId = nassert.getObjectId()
        let actual = getDefActual({ _id: objId })
        let expected = {
          _id: '_mock_'
        }
        test(actual, expected)
      })

      it('should pass assertion when actual is equal to expected with regexp field', () => {
        let objId = nassert.getObjectId()
        let actual = getDefActual({ _id: objId })
        let expected = {
          name: /^John/
        }
        test(actual, expected)
      })

      it('should pass assertion when actual is equal to expected with _mock_ field', () => {
        let objId = nassert.getObjectId()
        let actual = getDefActual({ _id: objId })
        let expected = {
          name: '_mock_'
        }
        test(actual, expected)
      })

      it('should pass assertion when actual is equal to expected', () => {
        let objId = nassert.getObjectId()
        let actual = getDefActual({ _id: objId })
        let expected = {
          _id: objId,
          __v: 'v', // can be any value
          createdAt: 'date', // can be any value
          updatedAt: 'date', // can be any value,
          name: 'John Smith'
        }
        test(actual, expected)
      })

      it('should pass assertion when actual is equal to expected (case with nested objects and arrays)', () => {
        let objId = nassert.getObjectId()
        let actual = {
          _id: objId,
          __v: 2,
          createdAt: new Date(2017, 12, 20),
          updatedAt: new Date(2017, 12, 21),
          name: 'John Smith',
          phones: [
            { type: 'mobile', number: 12345 },
            { type: 'work', number: 67890 }
          ],
          account: {
            accountNumber: '111111',
            created: new Date(2010, 11, 15)
          }
        }
        let expected = {
          _id: objId.toString(),
          __v: 'v', // can be any value
          createdAt: 'date', // can be any value
          updatedAt: 'date', // can be any value,
          name: 'John Smith',
          phones: [
            { type: 'mobile', number: '_mock_' }
          ],
          account: {
            accountNumber: '111111',
            created: '_mock_'
          }
        }
        test(actual, expected)
      })
    })
  })

  describe('assertResponse', () => {
    function test(res, expectedStatus, expectedBody, expectedErr) {
      try {
        nassert.assertResponse(res, expectedStatus, expectedBody)
        if (expectedErr) {
          throw new Error('Test must fail, because an error is expected')
        }
      } catch (err) {
        if (expectedErr instanceof Error) {
          should(err.message).eql(expectedErr.message)
        } else {
          throw err
        }
      }
    }

    it('should fail assertion when expectedStatus is 204 and res.headers.content-type is not undefined', () => {
      let res = {
        headers: { 'content-type': 'application/json' },
        body: {}
      }
      let expectedStatus = 204
      let expectedBody = undefined
      let expectedErr = new Error('expected \'application/json\' to be undefined')
      test(res, expectedStatus, expectedBody, expectedErr)
    })

    it('should fail assertion when expectedStatus is 204 and res.body is not undefined', () => {
      let res = {
        headers: {},
        body: { userId: 1 }
      }
      let expectedStatus = 204
      let expectedBody = undefined
      let expectedErr = new Error('expected Object { userId: 1 } to be empty')
      test(res, expectedStatus, expectedBody, expectedErr)
    })

    it('should pass assertion when expectedStatus is 204 and res.headers.content-type and res.body are empty', () => {
      let res = {
        headers: {},
        body: {}
      }
      let expectedStatus = 204
      let expectedBody = undefined
      test(res, expectedStatus, expectedBody)
    })

    it('should fail assertion when expectedStatus is not 204 and res.headers.content-type is not application/json', () => {
      let res = {
        headers: { 'content-type': 'application/xml' },
        body: { userId: 1 }
      }
      let expectedStatus = 201
      let expectedBody = { userId: 1 }
      let expectedErr = new Error('expected \'application/xml\' to match /application\\/json/')
      test(res, expectedStatus, expectedBody, expectedErr)
    })

    it('should fail assertion when expectedStatus is not 204 and res.body is not equal ro expectedBody', () => {
      let res = {
        headers: { 'content-type': 'application/json' },
        body: { userId: 1 }
      }
      let expectedStatus = 201
      let expectedBody = { userId: 2 }
      let expectedErr = new Error('expected 1 to equal 2 at path userId')
      test(res, expectedStatus, expectedBody, expectedErr)
    })

    it('should pass assertion when expectedStatus is not 204 and res.body is equal to expectedBody', () => {
      let res = {
        headers: { 'content-type': 'application/json' },
        body: { userId: 1 }
      }
      let expectedStatus = 201
      let expectedBody = { userId: 1 }
      test(res, expectedStatus, expectedBody)
    })
  })

  describe('sinonMatch', () => {
    function test(actualVal, expectedVal, expectedRes) {
      let match = nassert.sinonMatch(expectedVal)
      try {
        let actualRes = match.test(actualVal)
        should(actualRes).eql(expectedRes)
      } catch (err) {
        if (expectedRes instanceof Error) {
          should(err.message).eql(expectedRes.message)
        } else {
          throw err
        }
      }
    }

    it('should call sinon.match and throw error when assertion is failed', () => {
      let actualVal = 5
      let expectedVal = 10
      let expectedRes = new Error('sinon.match AssertionError: expected 5 to equal 10')
      test(actualVal, expectedVal, expectedRes)
    })

    it('should call sinon.match and return true when assertion is passed', () => {
      let actualVal = 10
      let expectedVal = 10
      let expectedRes = true
      test(actualVal, expectedVal, expectedRes)
    })
  })

  describe('assertFn', () => {
    let _inst = {
      doWork: () => {}
    }

    beforeEach(() => {
      sinon.stub(_inst, 'doWork')
    })

    afterEach(() => {
      _inst.doWork.restore()
    })

    function test({ params, expected }) {
      if (_.isError(expected)) {
        should(nassert.assertFn.bind(nassert, params)).throw(expected.message)
      } else {
        nassert.assertFn(params)
      }
    }

    function getDefaultParams(ex) {
      let params = {
        inst: _inst,
        fnName: 'doWork'
      }
      return _.extend(params, ex)
    }

    it('should throw error when service method called, but expectedArgs and expectedMultipleArgs are not provided', () => {
      let params = getDefaultParams()
      let expected = new Error('Expected that doWork wouldn\'t be called')

      _inst.doWork()

      test({ params, expected })
    })

    it('should throw error when service method called once, but callCount is 2', () => {
      let params = getDefaultParams({ callCount: 2, expectedArgs: 'p1' })
      let expected = new Error('Expected that doWork called 2 times')

      _inst.doWork()

      test({ params, expected })
    })

    it('should throw error when service method doesn\'t called, but expectedArgs is provided', () => {
      let params = getDefaultParams({ expectedArgs: 'p1' })
      let expected = new Error('Expected that doWork called once')

      test({ params, expected })
    })

    it('should throw error when service method doesn\'t called, but callCount is 2', () => {
      let params = getDefaultParams({ callCount: 2, expectedArgs: 'p1' })
      let expected = new Error('Expected that doWork called 2 times')

      test({ params, expected })
    })

    it('should throw error when service method called twice and expectedArgs is provided', () => {
      let params = getDefaultParams({ expectedArgs: 'p1' })
      let expected = new Error('Expected that doWork called once')

      _inst.doWork('p1')
      _inst.doWork('p1')

      test({ params, expected })
    })

    it('should throw error when service method called with one arg, but expectedArgs is `_without-args_`', () => {
      let params = getDefaultParams({ expectedArgs: '_without-args_' })
      let expected = new Error('Expected that doWork called without args')

      _inst.doWork('p1')

      test({ params, expected })
    })

    it('should throw error when service method called with [p1, p3], but expectedMultipleArgs is [p1, p2]', () => {
      let params = getDefaultParams({ expectedMultipleArgs: ['p1', 'p2'] })
      let expected = new Error('Expected that doWork called with multiple args')

      _inst.doWork('p1', 'p3')

      test({ params, expected })
    })

    it('should throw error when service method called with p1, but expectedArgs is p2', () => {
      let params = getDefaultParams({ expectedArgs: 'p2' })
      let expected = new Error(`sinon.match AssertionError: expected 'p1' to equal 'p2'`)

      _inst.doWork('p1')

      test({ params, expected })
    })

    it('shouldn\'t throw error when service method doesn\'t called and expectedArgs, expectedMultipleArgs are not provided', () => {
      let params = getDefaultParams()

      test({ params, expected: null })
    })

    it('shouldn\'t throw error when service method called without args and expectedArgs is `_without-args_`', () => {
      let params = getDefaultParams({ expectedArgs: '_without-args_' })

      _inst.doWork()

      test({ params, expected: null })
    })

    it('shouldn\'t throw error when service method called with [p1, p2] and expectedMultipleArgs is [p1, p2]', () => {
      let params = getDefaultParams({ expectedMultipleArgs: ['p1', 'p2'] })

      _inst.doWork('p1', 'p2')

      test({ params, expected: null })
    })

    it('shouldn\'t throw error when service method called with p1 and expectedArgs is p1', () => {
      let params = getDefaultParams({ expectedArgs: 'p1' })

      _inst.doWork('p1')

      test({ params, expected: null })
    })
  })
})


index.js
const _ = require('lodash')
const bson = require('bson')
const should = require('should')

exports.getObjectId = () => new bson.ObjectId()
exports.getObjectIdStr = exports.getObjectId().toString()

let _sinon
exports.initSinon = (sinon) => _sinon = sinon

/* eslint max-statements: off */
exports.assert = (actual, expected, isEqual) => {
  if (_assertIfExpectedIsNil(actual, expected) ||
      _assertIfExpectedIsSimplePrim(actual, expected) ||
      _assertIfExpectedIsArrayOfSimplePrim(actual, expected)
  ) {
    return
  }

  actual = _convertModelsToPlain(actual)

  let expectedPaths = _getObjectPaths(expected)
  if (isEqual) {
    let actualPaths = _getObjectPaths(actual)
    actualPaths = _.sortBy(actualPaths)
    expectedPaths = _.sortBy(expectedPaths)
    should(actualPaths).eql(expectedPaths)
  }

  let path
  try {
    for (let i = 0; i < expectedPaths.length; i++) {
      path = expectedPaths[i]
      let field = _getActualField(path)
      let actualVal = _.get(actual, path)
      let expectedVal = _.get(expected, path)
      _assert(field, actualVal, expectedVal)
    }
  } catch (err) {
    let newErr = new Error(`${err.message} at path ${path}`)
    throw newErr
  }
}

exports.assertResponse = (res, expectedStatus, expectedBody) => {
  if (expectedStatus === 204) {
    should(res.headers['content-type']).be.undefined()
    should(res.body).be.empty()
  } else {
    should(res.headers['content-type']).match(/application\/json/)
    exports.assert(res.body, expectedBody)
  }
}

exports.assertFn = ({ inst, fnName, callCount = 1, nCall = 0, expectedArgs, expectedMultipleArgs }) => {
  let assertion

  if (!expectedArgs && !expectedMultipleArgs) {
    assertion = inst[fnName].called
    should(assertion).equal(false, `Expected that ${fnName} wouldn't be called`)
    return
  }

  assertion = inst[fnName].callCount
  let calledMessage = callCount === 1 ? 'once' : `${callCount} times`
  should(assertion).equal(callCount, `Expected that ${fnName} called ${calledMessage}`)

  if (expectedArgs === '_without-args_') {
    assertion = inst[fnName].getCall(nCall).calledWithExactly()
    should(assertion).equal(true, `Expected that ${fnName} called without args`)
  } else if (expectedMultipleArgs) {
    assertion = inst[fnName].getCall(nCall).calledWithExactly(...expectedMultipleArgs)
    should(assertion).equal(true, `Expected that ${fnName} called with multiple args`)
  } else {
    assertion = inst[fnName].getCall(nCall).calledWithExactly(exports.sinonMatch(expectedArgs))
    should(assertion).equal(true, `Expected that ${fnName} called with single arg`)
  }
}

exports.sinonMatch = (expected) => {
  return _sinon.match(actual => {
    try {
      exports.assert(actual, expected)
      return true
    } catch (err) {
      let newErr = new Error('sinon.match AssertionError: ' + err.message)
      throw newErr
    }
  })
}

function _convertModelsToPlain(actual) {
  if (_isMongooseModel(actual)) {
    return actual.toObject()
  }

  if (_.isArray(actual)) {
    return _.map(actual, item => {
      if (_isMongooseModel(actual)) {
        return item.toObject()
      }
      return item
    })
  }

  return actual
}

function _assert(field, actual, expected) {
  if (field === '_id' || expected instanceof bson.ObjectId) {
    _assertObjectId(actual, expected)
  } else if (field === '__v') {
    should(actual).instanceOf(Number)
  } else if (field === 'createdAt') {
    should(actual).instanceOf(Date)
  } else if (field === 'updatedAt') {
    should(actual).instanceOf(Date)
  } else if (expected instanceof RegExp) {
    should(actual).match(expected)
  } else if (expected === '_mock_') {
    should(actual).be.ok()
  } else {
    exports.assert(actual, expected)
  }
}

function _assertObjectId(actual, expected) {
  if (expected === '_mock_') {
    should(_safeToString(actual)).match(/^[a-z|\d]{24}$/)
  } else {
    should(_safeToString(actual)).equal(_safeToString(expected))
  }
}

function _assertIfExpectedIsNil(actual, expected) {
  if (!_.isNil(expected)) {
    return false
  }
  should(actual).not.be.ok()
  return true
}

function _assertIfExpectedIsArrayOfSimplePrim(actual, expected) {
  if (!_.isArray(expected) || !_.every(expected, _isSimplePrim)) {
    return false
  }
  should(actual).eql(expected)
}

function _assertIfExpectedIsSimplePrim(actual, expected) {
  if (!_isSimplePrim(expected)) {
    return false
  }
  should(actual).eql(expected)
  return true
}

function _isSimplePrim(prim) {
  return _.isBoolean(prim) ||
         _.isNumber(prim) ||
         _.isString(prim) ||
         _.isDate(prim) ||
         _.isSymbol(prim) ||
         _.isRegExp(prim)
}

function _isMongooseModel(val) {
  return _.get(val, 'constructor.name') === 'model' && _.isFunction(val.toObject)
}

function _safeToString(val) {
  return _.isNil(val) ? val : val.toString()
}

function _getObjectPaths(item, curPath = '', isArray = false) {
  let paths = []
  _.each(item, (val, key) => {
    if (val instanceof bson.ObjectId) {
      val = _safeToString(val)
    }
    let newPath = isArray ? `${curPath}[${key}]` : `${curPath}.${key}`
    if (_isSimplePrim(val)) {
      paths.push(newPath)
    }
    if (_.isArray(val)) {
      paths = paths.concat(_getObjectPaths(val, newPath, true))
    } else if (_.isObject(val)) {
      paths = paths.concat(_getObjectPaths(val, newPath))
    }
  })
  if (!curPath) {
    paths = _.map(paths, path => _.trimStart(path, '.'))
  }
  return paths
}

function _getActualField(path) {
  let fields = path.split('.')
  return _.last(fields)
}
